### PR TITLE
Start implementing the federation server keys API.

### PR DESCRIPTION
**Pull request by NegativeMjark**
Updates the clientapi server to allow the server keys to be passed in the environment.
This means that the federationapi and clientapi can be set up with the same keys.
Both will generate random keys if no keys are supplied so that the existing integration tests will continue to run.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/federationapi/config/config.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	// The name of the server. This is usually the domain name, e.g 'matrix.org', 'localhost'.
+	ServerName gomatrixserverlib.ServerName
+	// The private key which will be used to sign requests.
+	PrivateKey ed25519.PrivateKey
+	// An arbitrary string used to uniquely identify the PrivateKey. Must start with the
+	// prefix "ed25519:".
+	KeyID gomatrixserverlib.KeyID
+	// A list of SHA256 TLS fingerprints for this server.
+	TLSFingerPrints []gomatrixserverlib.TLSFingerprint
+	// How long the keys are valid for.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/federationapi/config/config.go]: Which keys? Keys we get from servers or our own keys?

File: src/github.com/matrix-org/dendrite/federationapi/readers/keys.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+// LocalKeys returns the local keys for the server.
+// See https://matrix.org/docs/spec/server_server/unstable.html#publishing-keys
+func LocalKeys(req *http.Request, cfg config.FederationAPI) util.JSONResponse {
+	keys, err := localKeys(cfg, time.Now().Add(cfg.ValidityPeriod))
+	if err != nil {
+		return util.ErrorResponse(err)
+	}
+	return util.JSONResponse{JSON: keys}
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/federationapi/readers/keys.go]: Don't we also need a `Code`?

File: src/github.com/matrix-org/dendrite/federationapi/config/config.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/federationapi/config/config.go]: Commented cd8d9c4

File: src/github.com/matrix-org/dendrite/common/readkey.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"encoding/base64"
+	"fmt"
+	"github.com/matrix-org/gomatrixserverlib"
+	"golang.org/x/crypto/ed25519"
+	"io"
+	"strings"
+)
+
+// ReadKey reads a server's private ed25519 key.
+func ReadKey(key string) (gomatrixserverlib.KeyID, ed25519.PrivateKey, error) {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/common/readkey.go]: Please give the format of `key`, which afaict is:
```
ed25519:somekeyid $base-64-encoded-seed-bytes
```

Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/common/readkey.go]: Oh you did it after refreshing the page. Never mind then.

File: src/github.com/matrix-org/dendrite/federationapi/config/config.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	PrivateKey ed25519.PrivateKey
+	// An arbitrary string used to uniquely identify the PrivateKey. Must start with the
+	// prefix "ed25519:".
+	KeyID gomatrixserverlib.KeyID
+	// A list of SHA256 TLS fingerprints for this server.
+	TLSFingerPrints []gomatrixserverlib.TLSFingerprint
+	// How long a remote server can cache our server key for before requesting it again.
+	// Increasing this number will reduce the number of requests made by remote servers
+	// for our key, but increases the period a compromised key will be considered valid
+	// by remote servers.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/federationapi/config/config.go]: **Excellent** explanation :star2: thank you!

**Comment by Kegsay** [comment]
Other than https://github.com/matrix-org/dendrite/pull/112#discussion_r117479595 - LGTM

File: src/github.com/matrix-org/dendrite/federationapi/readers/keys.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/federationapi/readers/keys.go]: Fixed 0bb12d5

**Pull request merged**